Add from_iterator_dense function

bytesnake opened the pull request:
This PR adds a method to create a `CrcMatrix` from an iterator. The data should be dense and in major column format. Some flavours:
 * we could use the syntax (3, 3, &[ ... ]).into() instead of CrcMatrix::from_iterator_dense(3, 3, &[ ... ])
 * different name? I wasn't sure about the `_dense`
 * use `Cow` in borrow mode without copying

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/oxfordcontrol/osqp.rs?pullRequest=9) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/oxfordcontrol/osqp.rs?pullRequest=9) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/oxfordcontrol/osqp.rs?pullRequest=9) it.</sub>

ebarnard commented:
Thanks.

I think `from_iterator_dense` is fine as I want the name to convey the fact that the resulting `CscMatrix` is dense even if the iterator containes zeros.

I've added a couple of other comments inline.

ebarnard commented on src/csc.rs:
```diff
@@ -134,6 +134,24 @@ impl<'a> CscMatrix<'a> {
+    /// there are more efficient methods to construct the sparse matrix. Note that this copies all
+    /// elements of the matrix.
+    ///
+    pub fn from_iterator_dense<'b, I: IntoIterator<Item=&'b float>>(nrows: usize, ncols: usize, iter: I) -> CscMatrix<'static> {
```
Could you change this to take `Item=float`. It's just slightly more idiomatic.

bytesnake replied:
Ok, I modified this. The function will now only accept owned data.

ebarnard commented on src/csc.rs:
```diff
@@ -134,6 +134,24 @@ impl<'a> CscMatrix<'a> {
+            ncols,
+            indptr: (0..ncols+1).map(|i| i * nrows).collect::<Vec<usize>>().into(),
+            indices: iter::repeat(0..nrows).take(ncols).flatten().collect::<Vec<usize>>().into(),
+            data: iter.into_iter().cloned().collect::<Vec<float>>().into(),
```
Could you check that `data.len() == nrows * ncols` and panic if that invariant is not upheld.

bytesnake commented:
I think a single `from_iterator_dense` function is not sufficient, because it requires the data to be owned. Therefore I added a second function called `from_slice_dense` which constructs a `CrcMatrix` and stores a reference to the slice. I modified the test too and added some `assert_eq!`.

ebarnard commented:
My concern with `from_slice_dense` is that for an `nrows * ncols` matrix you must still allocate space for `(nrows + 1) * ncols` `usize`s in the `indices` and `indptr` vectors, but the name and lifetime on the method impl imply that no allocation takes place.

Moreover you end up with a `CscMatrix<'a>` where `'a` is the lifetime of the borrowed slice which makes it very hard to re-use the `'static` `indptr` and `indices` allocations.

I can't imagine a scenario where the cost of allocating `(nrows + 1) * ncols` elements is acceptable but allocating `(2 * nrows + 1) * ncols` elements is not.

Instead I'd accept something like this which allows creating a `CscMatrix` from a data slice and existing `CscMatrix` of the right shape.

```rust
impl CscMatrix<'mat> {
    fn with_data_slice(&'a self, data: &'data [float]) -> CscMatrix<'a> where 'mat: 'a, 'data: 'a {
        assert_eq!(data.len(), self.indices.len());
        ...
    }
}
```

bytesnake commented:
Yes I haven't thought about it, but this will then be as bad as the iterator constructor. I don't think that a `with_data_slice` will help me to simplify my example, but I don't mind using the iterator functions. Your changes are looking fine.

ebarnard commented:
Merged in 64ad54784e10c36a6180799cc9735789e285e14c
